// @sentinel/common/esm2022/utils/query-params-navigation-queue.service.mjs
import { Router } from '@angular/router';
import { inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class QueryParamsNavigationQueueService {
    constructor() {
        this.queue = Promise.resolve(true);
        this.router = inject(Router);
    }
    enqueue(queryParams) {
        const enqueue = async () => {
            try {
                await this.queue;
            }
            catch (error) {
                console.error(error);
            }
            return await this.router.navigate([], {
                queryParams: queryParams,
                queryParamsHandling: 'merge',
                replaceUrl: true,
            });
        };
        this.queue = enqueue();
        return this.queue;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QueryParamsNavigationQueueService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QueryParamsNavigationQueueService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QueryParamsNavigationQueueService, decorators: [{
            type: Injectable
        }] });
