// @sentinel/common/esm2022/utils/utils.mjs
export function isArray(x) {
    return Array.isArray(x);
}
export function hasValue(value) {
    return value !== undefined && value !== null;
}
export function propertyOf(name) {
    return name.toString();
}
export function trackById(index, item) {
    return item.id;
}
export function trackByIndex(index) {
    return index;
}
/**
 * Tries to match string value to one of the keys of enum. Returns undefined if it fails to match
 */
export const enumFromValue = (enumObj, val) => {
    const enumName = Object.keys(enumObj).find((k) => enumObj[k] === val);
    return !enumName ? undefined : enumObj[enumName];
};
export function toQueryParams(inputObject) {
    const params = {};
    const entries = Object.entries(inputObject);
    entries.forEach((entry) => (params[entry[0]] = entry[1]));
    return params;
}
export function isLoading(state, id) {
    if ('id' in state) {
        return state.isLoading && id === state.id;
    }
    if ('ids' in state) {
        return state.isLoading && state.ids.includes(id);
    }
    return false;
}
export function prettifyEnum(value) {
    if (value) {
        const lowercased = value.toLowerCase();
        const splitWords = lowercased.split('_');
        for (let i = 0; i < splitWords.length; i++) {
            splitWords[i] = splitWords[i].charAt(0).toUpperCase() + splitWords[i].slice(1);
        }
        return splitWords.join(' ');
    }
    else {
        return '';
    }
}
export function slugify(value) {
    if (value === null || value === undefined) {
        return undefined;
    }
    else {
        return encodeURIComponent(value);
    }
}
export function slugifyDate(value) {
    return value.valueOf();
}
export function deslugify(value) {
    if (value === null || value === undefined) {
        return undefined;
    }
    else {
        return decodeURIComponent(value);
    }
}
export function deslugifyTimestamp(value) {
    if (value === null || value === undefined) {
        return undefined;
    }
    else {
        const deslugified = deslugify(value);
        if (deslugified) {
            const timestamp = Number.parseInt(deslugified);
            if (!Number.isNaN(timestamp)) {
                return new Date(timestamp);
            }
        }
        return undefined;
    }
}
export function isDate(value) {
    return (value instanceof Date || (typeof value === 'object' && Object.prototype.toString.call(value) === '[object Date]'));
}
