// @sentinel/common/esm2022/graphql/api/graphql-context-util.mjs
import { HttpHeaders } from '@angular/common/http';
const IGNORE_ERROR_MESSAGE_ATTR = 'ignoreErrorMessage';
export function createApolloContextIgnoreErrorMessage(ignore) {
    return { [IGNORE_ERROR_MESSAGE_ATTR]: ignore };
}
export function retrieveApolloContextIgnoreErrorMessage(context) {
    if (context[IGNORE_ERROR_MESSAGE_ATTR] !== undefined) {
        return context[IGNORE_ERROR_MESSAGE_ATTR];
    }
    else {
        return false;
    }
}
const ERROR_TITLE_ATTR = 'errorTitle';
export function createApolloContextErrorTitle(title) {
    return { [ERROR_TITLE_ATTR]: title };
}
export function retrieveApolloContextErrorTitle(context) {
    if (context[ERROR_TITLE_ATTR] !== undefined) {
        return context[ERROR_TITLE_ATTR];
    }
    else {
        return 'Unknown operation';
    }
}
export const RETRYABLE_ATTR = 'retryable';
export function createApolloContextRetryable(value) {
    return { [RETRYABLE_ATTR]: value };
}
export function retrieveApolloContextRetryable(context) {
    if (context[RETRYABLE_ATTR] !== undefined) {
        return context[RETRYABLE_ATTR];
    }
    else {
        return false;
    }
}
export function wrapApolloContextForFileUpload(currContext) {
    currContext.useMultipart = true;
    if (currContext.headers) {
        currContext.headers = currContext.headers.append('Apollo-Require-Preflight', `true`);
    }
    else {
        currContext.headers = new HttpHeaders().set('Apollo-Require-Preflight', `true`);
    }
    return currContext;
}
