// @sentinel/common/esm2022/apollo-services/general/apollo-query.service.mjs
import { NetworkStatus } from '@apollo/client/core';
import { createApolloContextErrorTitle, createApolloContextIgnoreErrorMessage, createApolloContextRetryable, } from '@sentinel/common/graphql';
import { BehaviorSubject, EMPTY, of } from 'rxjs';
import { catchError, distinctUntilChanged, filter, map, shareReplay, switchMap, tap } from 'rxjs/operators';
const DEFAULT_OPTIONS = {
    fetchPolicy: 'network-only',
    pollInterval: undefined,
    retryable: true,
    ignoreErrorMessage: false,
};
export class ApolloQueryService {
    constructor(getQuery, operationName, options) {
        this.getQuery = getQuery;
        this.initialized$ = new BehaviorSubject(false);
        this.isInErrorState = false;
        this.operationName = operationName;
        this.options = options ?? DEFAULT_OPTIONS;
        this.data$ = this.initialized$.pipe(shareReplay({ refCount: true }), switchMap(() => this.queryRef?.valueChanges.pipe(
        // ignoring all responses not containing data
        filter((queryRes) => queryRes.data !== undefined && queryRes.networkStatus === NetworkStatus.ready), map((queryRes) => this.toData(queryRes)), catchError(() => EMPTY)) ?? EMPTY));
        this.isRefetching$ = this.initialized$.pipe(switchMap(() => this.queryRef?.valueChanges.pipe(map((queryRes) => (queryRes.networkStatus === NetworkStatus.refetch && queryRes.loading) ||
            queryRes.error !== undefined ||
            false), catchError(() => of(false)), distinctUntilChanged()) ?? EMPTY));
        this.isLoading$ = this.initialized$.pipe(switchMap(() => this.loadingInternal$ ?? EMPTY));
        this.hasError$ = this.initialized$.pipe(switchMap(() => this.hasErrorInternal$ ?? EMPTY));
    }
    recoverAfterError() {
        this.initialize();
    }
    initialize() {
        this.initQueryWatch();
        this.initLoading();
        this.initHasError();
        this.initialized$.next(true);
    }
    initQueryWatch() {
        this.queryRef = this.getQuery.watch(this.variables, {
            useInitialLoading: true,
            pollInterval: this.options?.pollInterval ?? DEFAULT_OPTIONS.pollInterval,
            fetchPolicy: this.options.fetchPolicy ?? DEFAULT_OPTIONS.fetchPolicy,
            nextFetchPolicy: 'network-only',
            errorPolicy: 'all',
            returnPartialData: true,
            notifyOnNetworkStatusChange: true, // to get all updates on loading state
            context: {
                ...createApolloContextErrorTitle(this.operationName),
                ...createApolloContextIgnoreErrorMessage(this.options?.ignoreErrorMessage !== undefined
                    ? this.options.ignoreErrorMessage
                    : DEFAULT_OPTIONS.ignoreErrorMessage ?? false),
                ...createApolloContextRetryable(this.options?.retryable !== undefined ? this.options?.retryable : DEFAULT_OPTIONS.retryable ?? true),
            },
        });
    }
    update(updateVariables) {
        this.variables = this.createVariables(updateVariables);
        if (!this.initialized$.getValue()) {
            this.initialize();
        }
        if (this.isInErrorState) {
            this.initialize();
        }
        if (this.queryRef && this.variables) {
            this.queryRef.setVariables(this.variables);
        }
    }
    refetch(updateVariables) {
        if (this.queryRef) {
            this.queryRef.options.context = {
                ...this.queryRef.options.context,
                ...createApolloContextIgnoreErrorMessage(true),
            };
        }
        this.variables = this.createVariables(updateVariables);
        this.queryRef?.refetch(this.variables);
    }
    initLoading() {
        if (this.queryRef) {
            this.loadingInternal$ = this.queryRef.valueChanges.pipe(
            // we ignore refetch loadings before it causes bad ux.
            filter((queryRes) => queryRes.networkStatus !== NetworkStatus.refetch), map((queryRes) => queryRes.loading || queryRes.error !== undefined), catchError(() => of(false)), distinctUntilChanged());
        }
    }
    initHasError() {
        if (this.queryRef) {
            const ref = this.queryRef;
            this.hasErrorInternal$ = ref.valueChanges.pipe(map((queryRes) => this.hasError(queryRes)), catchError(() => of(true)), distinctUntilChanged(), tap((hasError) => (this.isInErrorState = hasError)));
        }
    }
    hasError(queryRes) {
        return (queryRes.error !== undefined || queryRes.errors !== undefined || queryRes.networkStatus === NetworkStatus.error);
    }
}
