// @sentinel/common/esm2022/sort-services/sort.service.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import equal from 'fast-deep-equal/es6';
import { BehaviorSubject } from 'rxjs';
import { distinctUntilChanged, shareReplay } from 'rxjs/operators';
import { INIT_SORT_TOKEN } from './order-token';
import * as i0 from "@angular/core";
export class SortService {
    constructor(initSort) {
        this.initSort = initSort;
        this.INIT_SORT = initSort ? initSort : { sort: undefined, sortDir: undefined };
        this.sortSubject$ = new BehaviorSubject(this.INIT_SORT);
        this.sort$ = this.sortSubject$.asObservable().pipe(distinctUntilChanged((prev, curr) => equal(prev, curr)), shareReplay({ refCount: true }));
    }
    get() {
        return this.sortSubject$.getValue();
    }
    update(sort) {
        this.sortSubject$.next({ ...sort });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SortService, deps: [{ token: INIT_SORT_TOKEN, optional: true }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SortService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SortService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [INIT_SORT_TOKEN]
                }] }] });
