// @sentinel/common/esm2022/resize/sentinel-resize.directive.mjs
import { Directive, EventEmitter, Output } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
import * as i1 from "./sentinel-resize.service";
export class SentinelResizeDirective {
    constructor(service, { nativeElement }, destroyRef) {
        this.destroyRef = destroyRef;
        this.sentinelResize = new EventEmitter();
        service
            .observe(nativeElement, 0)
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe((event) => {
            this.sentinelResize.emit(event);
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelResizeDirective, deps: [{ token: i1.SentinelResizeService }, { token: i0.ElementRef }, { token: i0.DestroyRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.0", type: SentinelResizeDirective, isStandalone: true, selector: "[sentinelResize]", outputs: { sentinelResize: "sentinelResize" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelResizeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[sentinelResize]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i1.SentinelResizeService }, { type: i0.ElementRef }, { type: i0.DestroyRef }], propDecorators: { sentinelResize: [{
                type: Output
            }] } });
