// @sentinel/common/esm2022/resize/sentinel-resize.service.mjs
import { Injectable } from '@angular/core';
import { SentinelResizeManager } from './sentinel-resize-manager';
import * as i0 from "@angular/core";
export class SentinelResizeService {
    constructor() {
        this.windowResize$ = SentinelResizeService.manager.root.resize;
    }
    static { this.manager = new SentinelResizeManager(0); }
    observe(target, throttleBy = 90) {
        return SentinelResizeService.manager.observe(target).throttleBy(throttleBy);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelResizeService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelResizeService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelResizeService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
