// @sentinel/common/esm2022/dynamic-env/config-factory.mjs
import { InjectionToken } from '@angular/core';
/**
 * Injection token of the application's dynamic config. Can be useful for dynamic injection into other modules
 * Use as dependency for providing a config through injections token and configurableModuleFactory.
 */
export const APP_CONFIG = new InjectionToken('SentinelConfig');
const appConfigFactory = appConfigFactoryFunc;
/**
 * App config provider. Provides dynamic config at runtime using a factory function.
 * USAGE EXAMPLE:
 *  @NgModule({
 *     declarations: [AppComponent],
 *     imports: [BrowserModule,],
 *     providers: [appConfigProvider]
 *     bootstrap: [AppComponent]
 *   })
 *  export class AppModule {}
 */
export const appConfigProvider = { provide: APP_CONFIG, useFactory: appConfigFactory };
/**
 * Allows access to parts of a config dynamically, at runtime.
 * USAGE EXAMPLE:
 *  @NgModule({
 *   imports: [CommonModule, SentinelAuthModule.forRoot(null)],
 *   providers: [
 *     {
 *       provide: SentinelAuthConfig,
 *       useFactory: configurableModuleFactory<KypoConfig>('authConfig'),
 *       deps: [APP_CONFIG],
 *     },
 *   ],
 * })
 *  export class AuthModule {}
 */
export const configurableModuleFactory = configurableModuleFactoryFunc;
function configurableModuleFactoryFunc(prop) {
    return (sentinelConfig) => sentinelConfig[prop];
}
function appConfigFactoryFunc() {
    return window.sentinelAppConfig;
}
