// @sentinel/common/esm2022/validators/sentinel-validators.mjs
import { Validators } from '@angular/forms';
export class SentinelValidators {
    static { this.noWhitespace = (control) => {
        const isWhitespace = (control.value || '').trim().length === 0;
        const isValid = !isWhitespace;
        return isValid ? null : { whitespace: true };
    }; }
    static { this.minDate = (minDate) => {
        return (control) => {
            const date = new Date(control.value);
            if (minDate.getTime() < date.getTime()) {
                return null;
            }
            else {
                return { minDate: { value: control.value, expected: minDate } };
            }
        };
    }; }
    static { this.maxDate = (maxDate) => {
        return (control) => {
            const date = new Date(control.value);
            if (maxDate.getTime() > date.getTime()) {
                return null;
            }
            else {
                return { maxDate: { value: control.value, expected: maxDate } };
            }
        };
    }; }
    static { this.min = (min) => {
        return Validators.min(min);
    }; }
    static { this.max = (max) => {
        return Validators.max(max);
    }; }
    static { this.required = (control) => {
        return Validators.required(control);
    }; }
    static { this.requiredTrue = (control) => {
        return Validators.requiredTrue(control);
    }; }
    static { this.email = (control) => {
        return Validators.email(control);
    }; }
    static { this.minLength = (minLength) => {
        return Validators.minLength(minLength);
    }; }
    static { this.maxLength = (maxLength) => {
        return Validators.maxLength(maxLength);
    }; }
    static { this.pattern = (pattern) => {
        return Validators.pattern(pattern);
    }; }
    static { this.nullValidator = (control) => {
        return Validators.nullValidator(control);
    }; }
    static { this.compose = (validators) => {
        return Validators.compose(validators);
    }; }
    static { this.composeAsync = (validators) => {
        return Validators.composeAsync(validators);
    }; }
}
