// @sentinel/common/esm2022/pagination-services/pagination.service.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { distinctUntilChanged, shareReplay } from 'rxjs/operators';
import { INIT_PAGINATION_EVENT_BASE_TOKEN, PAGINATION_TYPE_TOKEN } from './pagination-token';
import equal from 'fast-deep-equal/es6';
import { CursorRequestedPaginationManager, isOffsetEvent, OffsetPaginationEvent, OffsetRequestedPaginationManager, isCursorEvent, CursorPaginationEvent, } from '@sentinel/common/pagination';
import * as i0 from "@angular/core";
import * as i1 from "./pagination-base-init";
export class PaginationService {
    constructor(initPaginationBase, paginationType) {
        this.initPaginationBase = initPaginationBase;
        this.paginationType = paginationType;
        if (!paginationType) {
            this.paginationType = 'cursor';
        }
        this.INIT_PAGINATION = createInitPagination(this.paginationType ?? 'cursor', this.initPaginationBase);
        this.paginationManager = createPaginationManager(this.INIT_PAGINATION.size, this.INIT_PAGINATION, this.paginationType ?? 'cursor');
        this.pagination$ = this.paginationManager.pagination$.asObservable().pipe(distinctUntilChanged((prev, curr) => equal(prev, curr)), shareReplay({ refCount: true }));
    }
    get() {
        return this.paginationManager.pagination$.getValue();
    }
    update(pagination) {
        this.paginationManager.update(pagination);
    }
    resetPosition() {
        this.paginationManager.resetPosition();
    }
    resetSize() {
        this.paginationManager.resetSize();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: PaginationService, deps: [{ token: INIT_PAGINATION_EVENT_BASE_TOKEN, optional: true }, { token: PAGINATION_TYPE_TOKEN, optional: true }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: PaginationService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: PaginationService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.PaginationBaseInit, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [INIT_PAGINATION_EVENT_BASE_TOKEN]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [PAGINATION_TYPE_TOKEN]
                }] }] });
export function createInitPagination(paginationType, initPaginationBase) {
    const paginationBase = initPaginationBase ?? { size: 10, sort: undefined, sortDir: undefined };
    if (paginationType === 'offset') {
        return new OffsetPaginationEvent(0, paginationBase.size, paginationBase.sort, paginationBase.sortDir);
    }
    else if (paginationType === 'cursor') {
        return new CursorPaginationEvent(null, paginationBase.size, paginationBase.sort, paginationBase.sortDir);
    }
    else {
        throw new Error(`PaginationType does not match supported values 'offset' and 'cursor'`);
    }
}
export function createPaginationManager(initSize, initPagination, paginationType) {
    if (paginationType === 'offset' && isOffsetEvent(initPagination)) {
        return new OffsetRequestedPaginationManager(initSize, initPagination);
    }
    else if (paginationType === 'cursor' && isCursorEvent(initPagination)) {
        return new CursorRequestedPaginationManager(initSize, initPagination);
    }
    else {
        throw new TypeError('Incompatible pagination type and initial pagination');
    }
}
