// @sentinel/common/esm2022/pagination-services/query-param-pagination.service.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { propertyOf, toQueryParams, } from '@sentinel/common/utils';
import { from } from 'rxjs';
import { concatMap, map, shareReplay, take, tap } from 'rxjs/operators';
import { INIT_PAGINATION_EVENT_BASE_TOKEN, MAX_PAGINATION_SIZE_TOKEN, PAGINATION_TYPE_TOKEN } from './pagination-token';
import { createPaginationManager, PaginationService } from './pagination.service';
import { CursorPaginationEvent, OffsetPaginationEvent } from '@sentinel/common/pagination';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@sentinel/common/utils";
import * as i3 from "./pagination-base-init";
export class QueryParamPaginationService extends PaginationService {
    constructor(activatedRoute, queryParamsQueueService, initPaginationBase, paginationType, initMaxPaginationSize) {
        super(initPaginationBase, paginationType);
        this.activatedRoute = activatedRoute;
        this.queryParamsQueueService = queryParamsQueueService;
        this.initPaginationBase = initPaginationBase;
        this.paginationType = paginationType;
        this.initMaxPaginationSize = initMaxPaginationSize;
        if (!paginationType) {
            this.paginationType = 'cursor';
        }
        this.maxPaginationSize = initMaxPaginationSize ? initMaxPaginationSize : 100;
        this.pagination$ = this.createPaginationObservable();
    }
    updateQueryParams(pagination) {
        return from(this.queryParamsQueueService.enqueue(toQueryParams(pagination)));
    }
    createPaginationObservable() {
        return this.fromRoute().pipe(tap((pagination) => (this.paginationManager = createPaginationManager(this.INIT_PAGINATION.size, pagination, this.paginationType ?? 'cursor'))), concatMap(() => this.paginationManager.pagination$), concatMap((pagination) => this.updateQueryParams(pagination).pipe(map(() => pagination))), shareReplay({ refCount: true }));
    }
    fromRoute() {
        return this.activatedRoute.queryParamMap.pipe(take(1), map((paramMap) => this.paginationFromParamMap(paramMap)));
    }
    paginationFromParamMap(paramMap) {
        let pagination;
        if (this.paginationType === 'offset') {
            pagination = this.offsetPaginationFromParamMap(paramMap);
        }
        else if (this.paginationType === 'cursor') {
            pagination = this.cursorPaginationFromParamMap(paramMap);
        }
        return pagination ?? this.INIT_PAGINATION;
    }
    offsetPaginationFromParamMap(paramMap) {
        const paginationBase = this.paginationBaseFromParamMap(paramMap);
        const pageStr = paramMap.get('page');
        if (paginationBase && pageStr !== null) {
            const parsedPage = Number.parseInt(pageStr);
            if (Number.isInteger(parsedPage)) {
                const page = Math.max(parsedPage, 0);
                return new OffsetPaginationEvent(page, paginationBase.size, paginationBase.sort, paginationBase.sortDir);
            }
        }
        return undefined;
    }
    cursorPaginationFromParamMap(paramMap) {
        const paginationBase = this.paginationBaseFromParamMap(paramMap);
        const cursor = paramMap.get('cursor');
        if (paginationBase) {
            return new CursorPaginationEvent(cursor, paginationBase.size, paginationBase.sort, paginationBase.sortDir);
        }
        return undefined;
    }
    paginationBaseFromParamMap(paramMap) {
        const sizeStr = paramMap.get(propertyOf('size'));
        const sort = paramMap.get(propertyOf('sort'));
        const sortDir = paramMap.get(propertyOf('sortDir'));
        if (sizeStr !== null) {
            const parsedSize = Number.parseInt(sizeStr);
            if (Number.isInteger(parsedSize)) {
                const positiveSize = Math.max(parsedSize, 1);
                const size = Math.min(positiveSize, this.maxPaginationSize);
                if (sort !== null && sortDir !== null) {
                    return { size: size, sort, sortDir: sortDir === 'asc' ? 'asc' : 'desc' };
                }
                else {
                    return { size: size, sort: undefined, sortDir: undefined };
                }
            }
        }
        return undefined;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QueryParamPaginationService, deps: [{ token: i1.ActivatedRoute }, { token: i2.QueryParamsNavigationQueueService }, { token: INIT_PAGINATION_EVENT_BASE_TOKEN, optional: true }, { token: PAGINATION_TYPE_TOKEN, optional: true }, { token: MAX_PAGINATION_SIZE_TOKEN, optional: true }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QueryParamPaginationService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: QueryParamPaginationService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.ActivatedRoute }, { type: i2.QueryParamsNavigationQueueService }, { type: i3.PaginationBaseInit, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [INIT_PAGINATION_EVENT_BASE_TOKEN]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [PAGINATION_TYPE_TOKEN]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [MAX_PAGINATION_SIZE_TOKEN]
                }] }] });
