// @sentinel/common/esm2022/model/forms/base-reactive-form.mjs
import { merge, Subject } from 'rxjs';
import { distinctUntilChanged, map, startWith } from 'rxjs/operators';
export class BaseReactiveForm {
    constructor(formGroup) {
        this.dirtySubject$ = new Subject();
        this.formGroup = formGroup;
        this.valid$ = this.formGroup.statusChanges.pipe(map(() => this.formGroup.valid), distinctUntilChanged(), startWith(this.formGroup.valid));
        const dirtyFromForm$ = this.formGroup.statusChanges.pipe(map(() => this.formGroup.dirty));
        this.dirty$ = merge(dirtyFromForm$, this.dirtySubject$).pipe(distinctUntilChanged(), startWith(this.dirty));
    }
    get valid() {
        return this.formGroup.valid;
    }
    get dirty() {
        return this.formGroup.dirty;
    }
    set dirty(value) {
        value ? this.formGroup.markAsDirty() : this.formGroup.markAsPristine();
        this.dirtySubject$.next(value);
    }
    set isSaving(value) {
        value ? this.formGroup.disable({ emitEvent: false }) : this.formGroup.enable({ emitEvent: false });
    }
}
