// @sentinel/common/esm2022/apollo-services/paginable/apollo-paginated-dmt.service.mjs
import { CursorPaginationConverter, CursorPaginationResolver, DmtBaseOrderingConverter, QueryCursorPaginationDataResolver, } from '@sentinel/common/graphql';
import { CursorPaginationEvent, isCursorEvent } from '@sentinel/common/pagination';
import { ApolloPaginatedBaseService } from '../general/apollo-paginated-base.service';
export class ApolloPaginatedDmtService extends ApolloPaginatedBaseService {
    constructor(getQuery, operationName, options) {
        super(getQuery, operationName, options);
        this.getQuery = getQuery;
    }
    queryToPaginationData(query) {
        return QueryCursorPaginationDataResolver.resolve(this.getQueryResponse(query));
    }
    toFragments(query) {
        return this.getQueryResponse(query).data;
    }
    createInitialPaginationEvent() {
        return new CursorPaginationEvent(null, 0, undefined, undefined);
    }
    createPaginationVariables(paginationEvent) {
        if (isCursorEvent(paginationEvent)) {
            return CursorPaginationConverter.convert(paginationEvent);
        }
        else {
            throw TypeError('Unsupported type of pagination');
        }
    }
    createOrderVariables(sort, dir) {
        if (sort && sort !== '' && dir && dir !== '') {
            return DmtBaseOrderingConverter.fromSingleSort(this.strToOrderable(sort), dir);
        }
        else {
            return {};
        }
    }
    createPaginationFromQueryResult(result) {
        if (isCursorEvent(this.lastPaginationEvent)) {
            const paginationData = this.queryToPaginationData(result.data);
            return CursorPaginationResolver.resolve(paginationData, this.lastPaginationEvent);
        }
        else {
            throw TypeError('Unsupported pagination');
        }
    }
}
