// @sentinel/common/fesm2022/sentinel-common-graphql.mjs
import { HttpHeaders } from '@angular/common/http';
import { of } from 'rxjs';
import { CursorPagination, OffsetPagination } from '@sentinel/common/pagination';

class ApolloResult {
    constructor(type, errors) {
        this.type = type;
        this.errors = errors;
    }
}

var ApolloResultType;
(function (ApolloResultType) {
    ApolloResultType[ApolloResultType["SUCCESS"] = 0] = "SUCCESS";
    ApolloResultType[ApolloResultType["ERROR"] = 1] = "ERROR";
})(ApolloResultType || (ApolloResultType = {}));

class DmtBaseOrderingConverter {
    static fromSingleSort(sort, sortDir) {
        if (sort && sortDir) {
            return { order: [{ field: sort, desc: sortDir === 'desc' }] };
        }
        else {
            return {};
        }
    }
}

class CursorPaginationConverter {
    static convert(pagination) {
        return {
            cursor: pagination.cursor ?? undefined,
            limit: pagination.size,
        };
    }
}
class OffsetPaginationConverter {
    static convert(pagination) {
        return {
            offset: pagination.page * pagination.size,
            limit: pagination.size,
        };
    }
}

class QueryCursorPaginationDataResolver {
    static resolve(cursorPaginatedResponse) {
        return {
            cursors: { prev: cursorPaginatedResponse.previous, next: cursorPaginatedResponse.next },
            elementsCount: cursorPaginatedResponse?.data?.length ?? 0,
            totalCount: cursorPaginatedResponse?.totalCount ?? 0,
        };
    }
}
class QueryOffsetPaginationDataResolver {
    static resolve(offsetPaginatedResponse) {
        return {
            elementsCount: offsetPaginatedResponse?.data?.length ?? 0,
            totalCount: offsetPaginatedResponse?.totalCount ?? 0,
        };
    }
}

const IGNORE_ERROR_MESSAGE_ATTR = 'ignoreErrorMessage';
function createApolloContextIgnoreErrorMessage(ignore) {
    return { [IGNORE_ERROR_MESSAGE_ATTR]: ignore };
}
function retrieveApolloContextIgnoreErrorMessage(context) {
    if (context[IGNORE_ERROR_MESSAGE_ATTR] !== undefined) {
        return context[IGNORE_ERROR_MESSAGE_ATTR];
    }
    else {
        return false;
    }
}
const ERROR_TITLE_ATTR = 'errorTitle';
function createApolloContextErrorTitle(title) {
    return { [ERROR_TITLE_ATTR]: title };
}
function retrieveApolloContextErrorTitle(context) {
    if (context[ERROR_TITLE_ATTR] !== undefined) {
        return context[ERROR_TITLE_ATTR];
    }
    else {
        return 'Unknown operation';
    }
}
const RETRYABLE_ATTR = 'retryable';
function createApolloContextRetryable(value) {
    return { [RETRYABLE_ATTR]: value };
}
function retrieveApolloContextRetryable(context) {
    if (context[RETRYABLE_ATTR] !== undefined) {
        return context[RETRYABLE_ATTR];
    }
    else {
        return false;
    }
}
function wrapApolloContextForFileUpload(currContext) {
    currContext.useMultipart = true;
    if (currContext.headers) {
        currContext.headers = currContext.headers.append('Apollo-Require-Preflight', `true`);
    }
    else {
        currContext.headers = new HttpHeaders().set('Apollo-Require-Preflight', `true`);
    }
    return currContext;
}

class ApolloOperationResultResolver {
    static resolve(fetchResult) {
        return fetchResult.errors
            ? new ApolloResult(ApolloResultType.ERROR, fetchResult.errors)
            : new ApolloResult(ApolloResultType.SUCCESS);
    }
    static handleError(err) {
        return of(new ApolloResult(ApolloResultType.ERROR, err));
    }
}

class CursorPaginationResolver {
    static resolve(paginationData, paginationEvent) {
        return new CursorPagination(paginationEvent.cursor, paginationData.elementsCount, paginationEvent.size, paginationData.totalCount, paginationData.cursors.next ?? null, paginationData.cursors.prev ?? null);
    }
}
class OffsetPaginationResolver {
    static resolve(paginationData, paginationEvent) {
        return new OffsetPagination(paginationEvent.page, paginationData.elementsCount, paginationEvent.size, paginationData.totalCount, Math.floor(paginationData.totalCount / paginationEvent.size));
    }
}

/**
 * Generated bundle index. Do not edit.
 */

export { ApolloOperationResultResolver, ApolloResult, ApolloResultType, CursorPaginationConverter, CursorPaginationResolver, DmtBaseOrderingConverter, OffsetPaginationConverter, OffsetPaginationResolver, QueryCursorPaginationDataResolver, QueryOffsetPaginationDataResolver, RETRYABLE_ATTR, createApolloContextErrorTitle, createApolloContextIgnoreErrorMessage, createApolloContextRetryable, retrieveApolloContextErrorTitle, retrieveApolloContextIgnoreErrorMessage, retrieveApolloContextRetryable, wrapApolloContextForFileUpload };
