// @sentinel/common/fesm2022/sentinel-common-pipes.mjs
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';
import { isArray, prettifyEnum, hasValue } from '@sentinel/common/utils';
import { AsyncPipe, CommonModule } from '@angular/common';

class SentinelHasElementsPipe {
    transform(value) {
        if (value === undefined || value === null) {
            return false;
        }
        if (isArray(value)) {
            return value.length > 0;
        }
        else {
            return value.data && value.data.length > 0;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelHasElementsPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelHasElementsPipe, name: "sentinelHasElements" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelHasElementsPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'sentinelHasElements',
                }]
        }] });

class SentinelPrettifyEnumPipe {
    transform(value) {
        return prettifyEnum(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelPrettifyEnumPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelPrettifyEnumPipe, name: "sentinelPrettifyEnum" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelPrettifyEnumPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'sentinelPrettifyEnum',
                }]
        }] });

/**
 * Uses async pipe internally but retypes null to undefined as this is more in line with TypeScript optional attributes
 * and results in less retyping and boilerplate if strict checking is active. It also allows to provide default value
 * for cases where undefined type is not possible/desired.
 */
class SentinelAsyncPipe extends AsyncPipe {
    transform(obj, defaultValue) {
        const asyncPipeResult = super.transform(obj);
        if (asyncPipeResult !== undefined && asyncPipeResult !== null) {
            return asyncPipeResult;
        }
        else {
            if (defaultValue) {
                return defaultValue;
            }
            else {
                return undefined;
            }
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelAsyncPipe, deps: null, target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelAsyncPipe, name: "sAsync", pure: false }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelAsyncPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'sAsync', pure: false }]
        }] });

class SentinelElementsTooltipPipe {
    transform(elements, name = 'element') {
        const lowerCased = name.toLowerCase();
        if (elements === undefined || elements.length === 0) {
            return `No ${lowerCased}s`;
        }
        if (elements.length === 1) {
            return `1 ${lowerCased}`;
        }
        return `${elements.length} ${lowerCased}s`;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelElementsTooltipPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelElementsTooltipPipe, name: "sentinelElementsTooltip" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelElementsTooltipPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'sentinelElementsTooltip' }]
        }] });

/**
 * Shortens string by slicing end of the string and replacing it with "..."
 */
class SentinelShortStringPipe {
    transform(text, maxLength = 10) {
        if (text === undefined || text === null) {
            return '';
        }
        return text.length > maxLength ? text.slice(0, maxLength) + '...' : text;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelShortStringPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelShortStringPipe, name: "sentinelShortString" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelShortStringPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'sentinelShortString',
                }]
        }] });

class SentinelTooltipOnShortenedPipe {
    transform(value, maxLength = 30) {
        if (value === undefined || value === null) {
            return '';
        }
        else {
            return value.length > maxLength ? value : '';
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelTooltipOnShortenedPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelTooltipOnShortenedPipe, name: "sentinelTooltipOnShortened" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelTooltipOnShortenedPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'sentinelTooltipOnShortened',
                }]
        }] });

class SentinelDashReplacePipe {
    transform(value, replaceWith = ' ') {
        return value.replace(new RegExp('[_-]'), replaceWith);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelDashReplacePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelDashReplacePipe, name: "sentinelDashReplace" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelDashReplacePipe, decorators: [{
            type: Pipe,
            args: [{ name: 'sentinelDashReplace' }]
        }] });

class SentinelDefaultValueStringPipe {
    transform(value, defaultValue = '') {
        return value ? value : defaultValue;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelDefaultValueStringPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelDefaultValueStringPipe, name: "sentinelDefaultValueString" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelDefaultValueStringPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'sentinelDefaultValueString' }]
        }] });

class SentinelDefaultValueArrayPipe {
    transform(value, defaultValue = []) {
        return value ? value : defaultValue;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelDefaultValueArrayPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelDefaultValueArrayPipe, name: "sentinelDefaultValueArray" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelDefaultValueArrayPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'sentinelDefaultValueArray',
                }]
        }] });

class SentinelHasValuePipe {
    transform(value) {
        return hasValue(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelHasValuePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelHasValuePipe, name: "sentinelHasValue" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelHasValuePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'sentinelHasValue',
                }]
        }] });

/**
 * Contains all custom pipes
 */
class SentinelPipesModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelPipesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelPipesModule, declarations: [SentinelDefaultValueArrayPipe,
            SentinelDefaultValueStringPipe,
            SentinelDashReplacePipe,
            SentinelHasValuePipe,
            SentinelTooltipOnShortenedPipe,
            SentinelShortStringPipe,
            SentinelAsyncPipe,
            SentinelElementsTooltipPipe,
            SentinelHasElementsPipe,
            SentinelPrettifyEnumPipe], imports: [CommonModule], exports: [SentinelDefaultValueArrayPipe,
            SentinelDefaultValueStringPipe,
            SentinelDashReplacePipe,
            SentinelTooltipOnShortenedPipe,
            SentinelShortStringPipe,
            SentinelAsyncPipe,
            SentinelElementsTooltipPipe,
            SentinelHasElementsPipe,
            SentinelPrettifyEnumPipe,
            SentinelHasValuePipe] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelPipesModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelPipesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        SentinelDefaultValueArrayPipe,
                        SentinelDefaultValueStringPipe,
                        SentinelDashReplacePipe,
                        SentinelHasValuePipe,
                        SentinelTooltipOnShortenedPipe,
                        SentinelShortStringPipe,
                        SentinelAsyncPipe,
                        SentinelElementsTooltipPipe,
                        SentinelHasElementsPipe,
                        SentinelPrettifyEnumPipe,
                    ],
                    exports: [
                        SentinelDefaultValueArrayPipe,
                        SentinelDefaultValueStringPipe,
                        SentinelDashReplacePipe,
                        SentinelTooltipOnShortenedPipe,
                        SentinelShortStringPipe,
                        SentinelAsyncPipe,
                        SentinelElementsTooltipPipe,
                        SentinelHasElementsPipe,
                        SentinelPrettifyEnumPipe,
                        SentinelHasValuePipe,
                    ],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { SentinelAsyncPipe, SentinelDashReplacePipe, SentinelDefaultValueArrayPipe, SentinelDefaultValueStringPipe, SentinelElementsTooltipPipe, SentinelHasElementsPipe, SentinelHasValuePipe, SentinelPipesModule, SentinelPrettifyEnumPipe, SentinelShortStringPipe, SentinelTooltipOnShortenedPipe };
