// @sentinel/common/esm2022/services/paginated-elements/paginated-elements.service.mjs
import { BehaviorSubject } from 'rxjs';
/**
 * Contains subjects and observables typically used when handling paginated data.
 * Subscribe to resource$ to receive latest data updates.
 */
export class PaginatedElementsService {
    constructor(pageSize = 10) {
        /**
         * True if server returned error response on the latest request, false otherwise
         * Change internally in extending service. Client should subscribe to the observable
         */
        this.hasErrorSubject$ = new BehaviorSubject(false);
        /**
         * True if server returned error response on the latest request, false otherwise
         * @contract must be updated every time new data are received
         */
        this.hasError$ = this.hasErrorSubject$.asObservable();
        /**
         * True if response to the latest request was not yet received
         * Change internally in extending service. Client should subscribe to the observable
         */
        this.isLoadingSubject$ = new BehaviorSubject(false);
        /**
         * True if response to the latest request was not yet received
         * @contract must be updated every time new data are received
         */
        this.isLoading$ = this.isLoadingSubject$.asObservable();
        this.resourceSubject$ = new BehaviorSubject(this.initSubject(pageSize));
        this.resource$ = this.resourceSubject$.asObservable();
    }
}
