// @sentinel/common/esm2022/services/resource-polling.service.mjs
import { ResourceService } from './resource.service';
import { merge, Subject, timer } from 'rxjs';
import { retryWhen, switchMap } from 'rxjs/operators';
/**
 * Adds polling behaviour to resource service
 */
export class ResourcePollingService extends ResourceService {
    constructor(pollPeriod = 5000) {
        super();
        /**
         * Observable triggering retry of polling after it was interrupted (e.g. by error)
         */
        this.retryPolling$ = new Subject();
        this.pollPeriod = pollPeriod;
        this.resource$ = merge(this.createPoll(), this.resourceSubject$.asObservable());
    }
    /**
     * Performs necessary operations and updates state of the service.
     * @contract Needs to update hasError subject and retry polling
     * if it was previously interrupted
     * @param params any other parameters required to update data in your concrete service
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    onManualResourceRefresh(...params) {
        if (this.hasErrorSubject$.getValue()) {
            this.retryPolling$.next(true);
        }
        this.hasErrorSubject$.next(false);
    }
    /**
     * Creates poll observable using a timer. You can extend the behaviour by piping the observable and applying
     * RxJs operators on it (e.g. takeWhile to stop polling on specific conditions)
     * @param dueTime optionally set a due time when the timer should start. By default it equals to the
     * poll period as it is expected that the first request is made from the service or component
     * using the service (to properly set up pagination and all other possible params like ids etc)
     */
    createPoll(dueTime = this.pollPeriod) {
        return timer(dueTime, this.pollPeriod).pipe(switchMap(() => this.refreshResource()), retryWhen(() => this.retryPolling$));
    }
}
