// @sentinel/common/esm2022/services/resource.service.mjs
import { BehaviorSubject } from 'rxjs';
/**
 * Holds resource state data in an observable.
 * Subscribe to resource$ to receive latest data updates.
 */
export class ResourceService {
    constructor() {
        /**
         * True if server returned error response on the latest request, false otherwise
         * Change internally in extending service. Client should subscribe to the observable
         */
        this.hasErrorSubject$ = new BehaviorSubject(false);
        /**
         * True if server returned error response on the latest request, false otherwise
         * @contract must be updated every time new data are received
         */
        this.hasError$ = this.hasErrorSubject$.asObservable();
        /**
         * True if response to the latest request was not yet received
         * Change internally in extending service. Client should subscribe to the observable
         */
        this.isLoadingSubject$ = new BehaviorSubject(false);
        /**
         * True if response to the latest request was not yet received
         * @contract must be updated every time new data are received
         */
        this.isLoading$ = this.isLoadingSubject$.asObservable();
        /**
         * Resource data of generic type. Change internally in extending service.
         * Client should subscribe to the observable
         * @contract must be updated every time new data are received
         */
        this.resourceSubject$ = new BehaviorSubject(undefined);
        /**
         * Resource data of generic type. Change internally in extending service.
         * Client should subscribe to the observable
         * @contract must be updated every time new data are received
         */
        this.resource$ = this.resourceSubject$.asObservable();
    }
}
