// @sentinel/common/esm2022/directives/s-let.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class SLetDirective {
    constructor(viewContainer, templateRef) {
        this.context = { sLet: null };
        viewContainer.createEmbeddedView(templateRef, this.context);
    }
    set sLet(value) {
        this.context.sLet = value;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SLetDirective, deps: [{ token: i0.ViewContainerRef }, { token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.0", type: SLetDirective, isStandalone: true, selector: "[sLet]", inputs: { sLet: "sLet" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SLetDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: '[sLet]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.ViewContainerRef }, { type: i0.TemplateRef }], propDecorators: { sLet: [{
                type: Input
            }] } });
