// @sentinel/common/esm2022/filter-services/filter.service.mjs
import equal from 'fast-deep-equal/es6';
import { BehaviorSubject } from 'rxjs';
import { distinctUntilChanged, map, shareReplay, tap } from 'rxjs/operators';
import { isArray } from '@sentinel/common/utils';
export class FilterService {
    constructor(ignoredKeys = [], ignoredActiveFilterKeys = ignoredKeys) {
        this.ignoredKeys = ignoredKeys;
        this.ignoredActiveFilterKeys = ignoredActiveFilterKeys;
        this.EMPTY_FILTER = this.createEmptyFilter();
        this.filterLocalStateSubject$ = new BehaviorSubject(this.EMPTY_FILTER);
        this.filterSubject$ = new BehaviorSubject(this.filterLocalStateSubject$.getValue());
        this.filterLocalState$ = this.filterLocalStateSubject$.asObservable().pipe(distinctUntilChanged((prev, curr) => equal(prev, curr)), shareReplay({ refCount: true }));
        this.filter$ = this.filterSubject$.asObservable().pipe(distinctUntilChanged((prev, curr) => equal(prev, curr)), tap((filter) => this.filterLocalStateSubject$.next(filter), shareReplay({ refCount: true })));
        this.isEmpty$ = this.filter$.pipe(map((filter) => this.areFilterKeysEqual(filter, this.EMPTY_FILTER)), distinctUntilChanged());
        this.activeFiltersCount$ = this.filterLocalState$.pipe(map((filter) => this.calculateActiveFiltersCount(filter)), distinctUntilChanged());
    }
    get() {
        return this.filterSubject$.getValue();
    }
    update(filter) {
        const curr = { ...this.filterLocalStateSubject$.getValue() };
        this.filterLocalStateSubject$.next({ ...curr, ...filter });
    }
    updateAndSubmit(filter) {
        this.update(filter);
        this.submit();
    }
    submit() {
        this.filterSubject$.next(this.filterLocalStateSubject$.getValue());
    }
    clear() {
        const currFilter = this.get();
        const emptyFilter = { ...this.EMPTY_FILTER };
        this.ignoredKeys.forEach((ignoredKey) => {
            if (currFilter[ignoredKey] !== undefined && currFilter[ignoredKey] !== null) {
                emptyFilter[ignoredKey] = currFilter[ignoredKey];
            }
        });
        this.filterLocalStateSubject$.next(emptyFilter);
        this.submit();
    }
    set(filter) {
        this.filterLocalStateSubject$.next(filter);
    }
    areFilterKeysEqual(prev, curr, compareIgnoredKeys = false) {
        const prevCopy = { ...prev };
        const currCopy = { ...curr };
        if (!compareIgnoredKeys) {
            this.ignoredKeys.forEach((ignoredKey) => {
                delete prevCopy[ignoredKey];
                delete currCopy[ignoredKey];
            });
        }
        return equal(currCopy, prevCopy);
    }
    calculateActiveFiltersCount(filter) {
        if (typeof filter === 'object') {
            let count = 0;
            const narrowedFilter = filter;
            Object.keys(narrowedFilter).forEach((key) => {
                if (!this.ignoredActiveFilterKeys.includes(key) &&
                    narrowedFilter[key] !== undefined &&
                    narrowedFilter[key] !== null &&
                    narrowedFilter[key] !== '' &&
                    (!isArray(narrowedFilter[key]) || narrowedFilter[key].length > 0)) {
                    count++;
                }
            });
            return count;
        }
        else {
            return 0;
        }
    }
}
