// @sentinel/common/esm2022/filter-services/query-param-filter.service.mjs
import { ActivatedRoute } from '@angular/router';
import { deslugify, deslugifyTimestamp, isArray, isDate, QueryParamsNavigationQueueService, slugify, slugifyDate, } from '@sentinel/common/utils';
import { combineLatest, from, merge } from 'rxjs';
import { concatMap, distinctUntilChanged, map, shareReplay, skip, tap } from 'rxjs/operators';
import { FilterService } from './filter.service';
import { inject } from '@angular/core';
export class QueryParamFilterService extends FilterService {
    constructor(ignoredKeys = [], ignoredActiveFilterKeys = ignoredKeys) {
        super(ignoredKeys, ignoredActiveFilterKeys);
        this.queryParamsQueueService = inject(QueryParamsNavigationQueueService);
        this.activatedRoute = inject(ActivatedRoute);
        this.filter$ = this.createFilterObservable();
    }
    toQueryParams(filter) {
        const params = {};
        const entries = Object.entries(filter);
        entries
            .filter((entry) => !this.ignoredKeys.map((key) => key.toString()).includes(entry[0]))
            .forEach((entry) => {
            if (isDate(entry[1])) {
                params[entry[0]] = this.slugifyDate(entry[1]);
            }
            else if (isArray(entry[1])) {
                params[entry[0]] = entry[1].map((entry) => this.slugify(entry?.toString()));
            }
            else {
                params[entry[0]] = this.slugify(entry[1]?.toString());
            }
        });
        return params;
    }
    updateQueryParams(filter) {
        return from(this.queryParamsQueueService.enqueue(this.toQueryParams(filter)));
    }
    slugify(value) {
        return slugify(value);
    }
    slugifyDate(value) {
        return slugifyDate(value);
    }
    deslugify(value) {
        return deslugify(value);
    }
    deslugifyTimestamp(value) {
        return deslugifyTimestamp(value);
    }
    getParamValue(paramMap, key) {
        return paramMap.get(key.toString());
    }
    createFilterObservable() {
        const fromRoute$ = this.fromRoute().pipe(tap((filter) => this.updateAndSubmit(filter)));
        const fromSubject$ = this.createFilterSubjectObservable();
        return merge(fromSubject$, fromRoute$).pipe(distinctUntilChanged((prev, curr) => this.areFilterKeysEqual(prev, curr, true)), concatMap((filter) => this.updateQueryParams(filter).pipe(map(() => filter))), shareReplay({ refCount: true }));
    }
    createFilterSubjectObservable() {
        // skip subject init from route
        return this.filterSubject$.asObservable().pipe(shareReplay({ refCount: true }), skip(1));
    }
    fromRoute() {
        return combineLatest([
            this.activatedRoute.queryParamMap,
            this.activatedRoute.paramMap,
            this.activatedRoute.data,
        ]).pipe(concatMap((routeAsyncAttrs) => this.createFilterFromRoute(routeAsyncAttrs[0], routeAsyncAttrs[1], routeAsyncAttrs[2])));
    }
}
