// @sentinel/common/esm2022/apollo-services/entity-modify/create-edit-entity.service.mjs
import { EMPTY } from 'rxjs';
import { concatMap, tap } from 'rxjs/operators';
import { EntityModifyBaseService } from './entity-modify-base.service';
export class CreateEditEntityService extends EntityModifyBaseService {
    constructor(entityName, notificationService) {
        super(entityName, notificationService);
        this.notificationService = notificationService;
        this.editMode = false;
    }
    createForm() {
        return this.getEntityId().pipe(tap((id) => (this.editMode = id !== undefined)), concatMap((id) => this.createFormFromEntityId(id)));
    }
    save() {
        return this.form$.pipe(tap(() => this.isSavingSubject$.next(true)), concatMap((form) => (form.valid ? this.createOrUpdateEntity(form) : EMPTY)), tap({
            next: () => this.isSavingSubject$.next(false),
            error: () => this.isSavingSubject$.next(false),
        }));
    }
    createOrUpdateEntity(form) {
        const save$ = this.editMode ? this.updateEntity(form) : this.createEntity(form);
        return save$.pipe(concatMap((result) => this.handleSaveResult(form, result)));
    }
    createSuccessSaveMessage() {
        return this.editMode ? `${this.entityName} was updated` : `${this.entityName} was created`;
    }
    createErrorSaveMessage() {
        const lowerCasedEntity = this.entityName.toLowerCase();
        return this.editMode ? `Updating ${lowerCasedEntity}` : `Creating ${lowerCasedEntity}`;
    }
}
