// @sentinel/common/esm2022/apollo-services/entity-modify/create-entity.service.mjs
import { EntityModifyBaseService } from './entity-modify-base.service';
import { of } from 'rxjs';
import { concatMap, tap } from 'rxjs/operators';
export class CreateEntityService extends EntityModifyBaseService {
    constructor(entityName, notificationService) {
        super(entityName, notificationService);
        this.notificationService = notificationService;
    }
    save() {
        return this.form$.pipe(concatMap((form) => (form.valid ? this.validFormSave() : of(false))));
    }
    createSuccessSaveMessage() {
        return `${this.entityName} was created`;
    }
    createErrorSaveMessage() {
        const lowerCasedEntity = this.entityName.toLowerCase();
        return `Creating ${lowerCasedEntity}`;
    }
    validFormSave() {
        return this.form$.pipe(tap(() => this.isSavingSubject$.next(true)), concatMap((form) => this.createEntity(form).pipe(concatMap((result) => this.handleSaveResult(form, result)))), tap({
            next: () => this.isSavingSubject$.next(false),
            error: () => this.isSavingSubject$.next(false),
        }));
    }
}
