// @sentinel/common/esm2022/apollo-services/entity-modify/edit-entity.service.mjs
import { of } from 'rxjs';
import { concatMap, tap } from 'rxjs/operators';
import { EntityModifyBaseService } from './entity-modify-base.service';
export class EditEntityService extends EntityModifyBaseService {
    constructor(entityName, notificationService) {
        super(entityName, notificationService);
        this.notificationService = notificationService;
    }
    save() {
        return this.form$.pipe(concatMap((form) => (form.valid ? this.validFormSave(form) : of(false))));
    }
    createSuccessSaveMessage() {
        return `${this.entityName} was updated`;
    }
    createErrorSaveMessage() {
        const lowerCasedEntity = this.entityName.toLowerCase();
        return `Updating ${lowerCasedEntity}`;
    }
    validFormSave(form) {
        this.isSavingSubject$.next(true);
        return this.updateEntity(form).pipe(concatMap((result) => this.handleSaveResult(form, result)), tap({
            next: () => this.isSavingSubject$.next(false),
            error: () => this.isSavingSubject$.next(false),
        }));
    }
}
