// @sentinel/common/esm2022/apollo-services/entity-modify/entity-modify-base.service.mjs
import { ActivatedRoute } from '@angular/router';
import { ApolloResultType } from '@sentinel/common/graphql';
import { BehaviorSubject } from 'rxjs';
import { concatMap, distinctUntilChanged, map, shareReplay, switchMap, tap } from 'rxjs/operators';
import { inject } from '@angular/core';
export class EntityModifyBaseService {
    constructor(entityName, notificationService) {
        this.notificationService = notificationService;
        this.isSavingSubject$ = new BehaviorSubject(false);
        this.isSaving$ = this.isSavingSubject$.asObservable().pipe(distinctUntilChanged());
        this.isLoadingSubject$ = new BehaviorSubject(true);
        this.isLoading$ = this.isLoadingSubject$.asObservable();
        this.hasErrorSubject$ = new BehaviorSubject(false);
        this.hasError$ = this.hasErrorSubject$.asObservable();
        this.activatedRoute = inject(ActivatedRoute);
        this.entityName = entityName;
        this.form$ = this.createForm().pipe(tap({
            next: () => this.isLoadingSubject$.next(false),
            error: () => {
                this.isLoadingSubject$.next(false);
                this.hasErrorSubject$.next(true);
            },
        }), shareReplay({ refCount: true, bufferSize: 1 }));
        this.canDeactivate$ = this.form$.pipe(switchMap((form) => form.dirty$), map((dirty) => !dirty), distinctUntilChanged());
    }
    cancel() {
        return this.form$.pipe(tap((form) => (form.dirty = false)), concatMap(() => this.handleCancel()));
    }
    handleSaveResult(form, result) {
        if (result.type === ApolloResultType.SUCCESS) {
            form.dirty = false;
            this.notificationService.emit('success', this.createSuccessSaveMessage());
            return this.onSaveSuccess();
        }
        else {
            return this.onSaveError();
        }
    }
}
