// @sentinel/common/esm2022/pipes/sentinel-pipes.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { SentinelShortStringPipe } from './sentinel-short-string.pipe';
import { SentinelAsyncPipe } from './sentinel-async.pipe';
import { SentinelTooltipOnShortenedPipe } from './sentinel-tooltip-on-shortened.pipe';
import { SentinelDashReplacePipe } from './sentinel-dash-replace.pipe';
import { SentinelDefaultValueStringPipe } from './sentinel-default-value-string.pipe';
import { SentinelDefaultValueArrayPipe } from './sentinel-default-value-array.pipe';
import { SentinelHasElementsPipe } from './has-elements.pipe';
import { SentinelPrettifyEnumPipe } from './prettify-enum.pipe';
import { SentinelElementsTooltipPipe } from './elements-tooltip.pipe';
import { SentinelHasValuePipe } from './sentinel-has-value.pipe';
import * as i0 from "@angular/core";
/**
 * Contains all custom pipes
 */
export class SentinelPipesModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelPipesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.0", ngImport: i0, type: SentinelPipesModule, declarations: [SentinelDefaultValueArrayPipe,
            SentinelDefaultValueStringPipe,
            SentinelDashReplacePipe,
            SentinelHasValuePipe,
            SentinelTooltipOnShortenedPipe,
            SentinelShortStringPipe,
            SentinelAsyncPipe,
            SentinelElementsTooltipPipe,
            SentinelHasElementsPipe,
            SentinelPrettifyEnumPipe], imports: [CommonModule], exports: [SentinelDefaultValueArrayPipe,
            SentinelDefaultValueStringPipe,
            SentinelDashReplacePipe,
            SentinelTooltipOnShortenedPipe,
            SentinelShortStringPipe,
            SentinelAsyncPipe,
            SentinelElementsTooltipPipe,
            SentinelHasElementsPipe,
            SentinelPrettifyEnumPipe,
            SentinelHasValuePipe] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelPipesModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.0", ngImport: i0, type: SentinelPipesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        SentinelDefaultValueArrayPipe,
                        SentinelDefaultValueStringPipe,
                        SentinelDashReplacePipe,
                        SentinelHasValuePipe,
                        SentinelTooltipOnShortenedPipe,
                        SentinelShortStringPipe,
                        SentinelAsyncPipe,
                        SentinelElementsTooltipPipe,
                        SentinelHasElementsPipe,
                        SentinelPrettifyEnumPipe,
                    ],
                    exports: [
                        SentinelDefaultValueArrayPipe,
                        SentinelDefaultValueStringPipe,
                        SentinelDashReplacePipe,
                        SentinelTooltipOnShortenedPipe,
                        SentinelShortStringPipe,
                        SentinelAsyncPipe,
                        SentinelElementsTooltipPipe,
                        SentinelHasElementsPipe,
                        SentinelPrettifyEnumPipe,
                        SentinelHasValuePipe,
                    ],
                }]
        }] });
