// @sentinel/common/esm2022/pagination/manager/requested-pagination-manager.mjs
import { BehaviorSubject } from 'rxjs';
export class RequestedPaginationManager {
    constructor(pageSize, initial) {
        this.defaultPageSize = pageSize;
        const event = initial ?? this.createInitialPagination();
        this.pagination$ = new BehaviorSubject(event);
    }
    update(pagination) {
        if (this.hasSortingChanged(pagination) || this.hasPageSizeChanged(pagination)) {
            this.resetPosition(pagination);
        }
        else {
            this.pagination$.next({ ...pagination });
        }
    }
    resetSize() {
        const curr = this.pagination$.getValue();
        if (curr) {
            curr.size = this.defaultPageSize;
            this.pagination$.next({ ...curr });
        }
    }
    hasSortingChanged(next) {
        const curr = this.pagination$.getValue();
        if (!curr) {
            return false;
        }
        const isDifferent = next.sort !== curr.sort || next.sortDir !== curr.sortDir;
        if (isDifferent) {
            return (!this.isDifferentEmptyValues(next.sort, curr.sort) || !this.isDifferentEmptyValues(next.sortDir, curr.sortDir));
        }
        return false;
    }
    hasPageSizeChanged(pagination) {
        const curr = this.pagination$.getValue();
        if (curr) {
            return pagination.size !== curr.size;
        }
        else {
            return false;
        }
    }
    isDifferentEmptyValues(valueA, valueB) {
        return ((valueA === undefined || valueA === null || valueA === '') &&
            (valueB === undefined || valueB === null || valueB === ''));
    }
}
